File: JULY-2025/31-07-2025/p898.js
/**
 * @param {number[]} arr
 * @return {number}
 */
var subarrayBitwiseORs = function (a) {
    let uniqueBitwiseOrs = new Set();
    let prevBitwiseOrs = new Set();
    for (const e of a) {
        const currOrs = new Set();
        for (const por of prevBitwiseOrs) currOrs.add(por | e);
        currOrs.add(e);
        for (let e of currOrs) uniqueBitwiseOrs.add(e);
        prevBitwiseOrs = currOrs;
    }
    return uniqueBitwiseOrs.size;
};

File: JULY-2025/31-07-2025/p898.rb
# @param {Integer[]} arr
# @return {Integer}
def subarray_bitwise_o_rs(arr)
  res = {}
  cur = {}
  
  arr.each do |x|
    nxt = {x => true}
    cur.each_key { |y| nxt[y | x] = true }
    cur.each_key { |k| res[k] = true }
    res[x] = true
    cur = nxt
  end
  
  cur.each_key { |k| res[k] = true }
  res.size
end